### ISSUE TITLE
Create qbank_filter plugin

### ISSUE THREAD

**Issue by mattporritt** [catalyst/moodle-MDL-70329#47]
Create a qbank plugin that allows searching for question bank question using filters.
This is to support this user story: https://docs.google.com/document/d/1JPBTqpCbtdRFTJuX-GIaxKyxtFLoUdPO/edit#heading=h.9c48xo1xbf51

We should base this on the existing Moodle core participant filter and extend the existing core functionality to support filter by OR criteria. The current implementation seems to only support AND.

Initial high level tasks:
* [ ] create the qbank_filter plugin to implement question bank filtering
* [ ] Add support to existing core filter library to support OR for groups of filters
* [ ] Create a new Moodle tracker under the MDL-70329 epic to apply the patch to the filtering library to core
* [ ] Add the updated filter library to the qbank_filter plugin
* [ ] Create a new Moodle tracker under the MDL-70329 epic to add the qbank_filter plugin to core

**Comment by mattporritt** [comment]
A question I have for the team is around filtering and AND versus OR
I want to confirm Moodle's exiting core functionality VS what is required in the user stories.  Originally, I had thought Moodle doesn't support "OR" logic, only "AND" logic. I was mistaken, instead it works differently to the way I thought.

In core right now the AND logic (as it relates to the screenshot below) works like:
* Every individual filter (Keyword, Roles and Enrolment methods) must have at least one match for the filter to be satisfied
* HOWEVER, individual filters can be set to ANY/OR. See "Roles" in the screen shot can be "student" OR "guest"
* Individual filters can also have ALL/AND. So in the case of tags (yet to be implemented) every tag must match
![participant and](https://user-images.githubusercontent.com/7369127/120143702-f2ea0900-c223-11eb-944e-ba215647fdb7.png)

In core right now OR logic works like:
* Only one individual filter (Keyword, Roles and Enrolment methods) must have at least one match for the filter to be satisfied
* Same logic for individual filters applies as for the AND logic
![participant or](https://user-images.githubusercontent.com/7369127/120143939-6724ac80-c224-11eb-9fb9-b501d8272308.png)

I had originally thought that we would need to add filtering so we could have "groups" of OR conditions. I've made a mock-up of what this would look like below. However, after re-reading the user stories I'm not sure this is required. I also think this may end up being confusing for users.
![participant group or](https://user-images.githubusercontent.com/7369127/120144185-d8fcf600-c224-11eb-86be-e7daed4f52bf.png)

My questions are:
* Can I confirm all of the user stories are covered by the current Moodle core filter behavior?
* Do we need to include the "grouped" OR conditions?

**Comment by abonaccorso** [comment]
We will decide on this in one weeks time as the question was raised in the community forum only last Friday.

**Comment by abonaccorso** [comment]
Based on the answers of the community we decided not to make the OR groups filter public in this project. Most community members who spoke up consider the groups OR filter a feature they would like to have in the future but that is not urgent for a first release. So we decided to develop the groups OR filter behind the scenes, so it could be released later on (UI would still need to be developed) but not to make it public for this project.
